### ISSUE TITLE
document importance of middleware order in compose call

### ISSUE THREAD

**Issue by wichert** [faceyspacey/redux-first-router#43]
I just spent a lovely bit of time trying to figure out why actions I dispatched from onAfterChange did not do anything. It turns out that the order of your middlewares in the compose call is very important. Mine setup code like this:

```
    const middlewares = [sagaMiddleware, beaconMiddleware, routeMiddleware] 
    const middlewareEnhancer = applyMiddleware(...middlewares)
    const storeEnhancers = [middlewareEnhancer, routeEnhancer]
    const composedEnhancer = composeWithDevTools(...storeEnhancers)
```

And the action I dispatched is supposed to be handled by the beacon middleware. What happened is that the dispatch function passed to onAfterChange is not the store dispatch, but a router-specific one that inserts the action at the point of the router middleware in the processing pipeline. In my case that meant no middleware would ever see my action since the router middleware was already the last one.

I am not entirely sure why the dispatch doesn't happen at the root level - I am guessing there is a good reason? It think it would be useful to document the importance of the middleware ordering to save others the time I spent figuring this out.

**Comment by faceyspacey** [comment]
Definitely. I've been meaning to do that. Basically you want to put the router as early as possible in the chain.  This is a typical thing that occurs with middleware. You usually end up with one middleware more important than the others that u want first. Feel free to provide a pr or i will get to it soon.

**Comment by faceyspacey** [comment]
The reason is because the middleware has no access to the final dispatch function. It only gets the "next" function where it's at. There could be a way to hack it through assign variables in the enhancer to the outer closure which the middleware also has access to. I don't like that approach--but if anyone has an idiom for an idiomatic way to repeat the entire middleware chain, I'm more than open to it. That said, I think it's a common occurrence (ie where order becomes a problem) cuz it can't correctly be done.

**Comment by faceyspacey** [comment]
U just gotta be aware of order with ur middlewares. Order matters when middleware need to create additional actions. Order matters when one middleware's work changes based on the result of a precious middleware.

**Comment by faceyspacey** [comment]
Ok, so I did what I mentioned about "hacking it through assigning variables in the enhancer." It's only a hack in terms of purity, but makes sense since this package has an enhancer, not just middleware. 

> For packages that have just the middleware, they don't have this luxury, in which case order matters.

There will always be some dependence on the order of middlewares, but in this case I was able to solve it *(since what we're talking about is dispatching an additional action)*. So no longer do we recreate the middleware pipeline--rather `_store` is assigned in an outer closure when the enhancer is created.

These changes will appear in the `@next` dist-tag on NPM in an hour or 2 along with query params support. Probably by next week they will be promoted from the `next` dist-tag.
